Predict Used Cars Price

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats
import seaborn as sns
import numpy as np

EDA

Load Dataset, examine its summary and structure

In [2]:
df = pd.read_csv('data_car.csv')
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [3]:
# summary statistics
df.describe()

Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549.000000  2.100000e+04  
50%       26.000000     18.000000   1385.000000  2.999500e+04  
75%       30.000000     22.000000   2009.000000  4.223125e+04  
max      354.000000    137.000000   5657.000000  2.065902e+06

In [4]:
print(df.shape) # Check number of rows and columns

(11914, 16)


There are 16 columns and 11914 rows

In [5]:
print(df.columns) # Check column names

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')


In [6]:
# Check data types and non-null values
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

8 categorical and 8 numerical values

In [7]:
#unique val in categorical cols
print(df['Make'].unique())
print(df['Model'].unique())
print(df['Engine Fuel Type'].unique())
print(df['Transmission Type'].unique())
print(df['Driven_Wheels'].unique())
print(df['Market Category'].unique())
print(df['Vehicle Size'].unique())
print(df['Vehicle Style'].unique())

['BMW' 'Audi' 'FIAT' 'Mercedes-Benz' 'Chrysler' 'Nissan' 'Volvo' 'Mazda'
 'Mitsubishi' 'Ferrari' 'Alfa Romeo' 'Toyota' 'McLaren' 'Maybach'
 'Pontiac' 'Porsche' 'Saab' 'GMC' 'Hyundai' 'Plymouth' 'Honda'
 'Oldsmobile' 'Suzuki' 'Ford' 'Cadillac' 'Kia' 'Bentley' 'Chevrolet'
 'Dodge' 'Lamborghini' 'Lincoln' 'Subaru' 'Volkswagen' 'Spyker' 'Buick'
 'Acura' 'Rolls-Royce' 'Maserati' 'Lexus' 'Aston Martin' 'Land Rover'
 'Lotus' 'Infiniti' 'Scion' 'Genesis' 'HUMMER' 'Tesla' 'Bugatti']
['1 Series M' '1 Series' '100' '124 Spider' '190-Class' '2 Series' '200'
 '200SX' '240SX' '240' '2' '3 Series Gran Turismo' '3 Series' '300-Class'
 '3000GT' '300' '300M' '300ZX' '323' '350-Class' '350Z' '360' '370Z' '3'
 '4 Series Gran Coupe' '4 Series' '400-Class' '420-Class' '456M'
 '458 Italia' '4C' '4Runner' '5 Series Gran Turismo' '5 Series'
 '500-Class' '500e' '500' '500L' '500X' '550' '560-Class' '570S' '575M'
 '57' '599' '5' '6 Series Gran Coupe' '6 Series' '600-Class' '6000'
 '612 Scaglietti' '626' '62' '65

Pre PRocessing

Check null/missing values, handle them, data transformation, feature engineering

In [8]:
#checking null values in each column
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

some columns has missing values - 3 in Engine fuel type column, 
69 in Engine HP, 30 in Engine Cylinders, 6 in number of doors and 3742 in market category

In [9]:
#handling missing value
df['Engine Fuel Type'].fillna(df['Engine Fuel Type'].mode()[0], inplace=True)

Only 3 missing value in above column and since its categorical used mode to fill them

In [10]:
df['Engine HP'].fillna(df['Engine HP'].mean(), inplace=True)

This column represents the horsepower of the engine, which is a numeric variable. So filled the missing values with the mean of the column.

In [11]:
df['Engine Cylinders'].fillna(df['Engine Cylinders'].mean(), inplace=True)

Represents the number of cylinders in the engine and is also a numeric variable. The missing values filled with the mean of the column.

In [12]:
df['Number of Doors'].fillna(df['Number of Doors'].mode()[0], inplace=True)

The number of doors in the car, which is a categorical variable with ordinal relationships.Filled the values with the mode.

 Market category has 3k missing values and it is a categorical variable, so choosing to fill it with 'unknown'

In [13]:
df['Market Category'].fillna('Unknown', inplace=True)

In [14]:
##rechecking nulll values
df.isna().sum()

Make                 0
Model                0
Year                 0
Engine Fuel Type     0
Engine HP            0
Engine Cylinders     0
Transmission Type    0
Driven_Wheels        0
Number of Doors      0
Market Category      0
Vehicle Size         0
Vehicle Style        0
highway MPG          0
city mpg             0
Popularity           0
MSRP                 0
dtype: int64

In [15]:
#dropping irrelevant column like popularity
df = df.drop(['Popularity'], axis=1)

In [16]:
from sklearn.preprocessing import LabelEncoder, StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split

In [17]:
# Convert categorical variables into numerical representations using LabelEncoder and OneHotEncoder
categorical_cols = ['Make', 'Model', 'Engine Fuel Type', 'Transmission Type', 'Driven_Wheels', 'Market Category', 'Vehicle Size', 'Vehicle Style']
ordinal_cols = ['Number of Doors']

In [18]:
#label encoder applying for ordinal columns
le = LabelEncoder()
df[ordinal_cols] = df[ordinal_cols].apply(le.fit_transform)

In [19]:
#onehot encoder for categorical columns
oh = OneHotEncoder(sparse=False, handle_unknown='ignore')
transformer = ColumnTransformer(transformers=[('onehot', oh, categorical_cols)], remainder='passthrough')
df_encoded = transformer.fit_transform(df)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [20]:
# Separate the target variable (MSRP) from the input features
X = df_encoded[:, :-1]
y = df_encoded[:, -1]

In [21]:
#feature scaling using StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [22]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

Modelling

In [23]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.model_selection import cross_val_score

In [24]:
# 3 different models using to check which one is better option
models = [
    RandomForestRegressor(random_state=42),
    GradientBoostingRegressor(random_state=42),
    SVR()
]

# cross-validation and evaluate the models by evaluation matrix
for model in models:
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    
    print("Model:", type(model).__name__)
    print("Mean Squared Error (MSE):", mse)
    print("Mean Absolute Error (MAE):", mae)
    print()

Model: RandomForestRegressor
Mean Squared Error (MSE): 84353302.26416208
Mean Absolute Error (MAE): 3098.007804477495

Model: GradientBoostingRegressor
Mean Squared Error (MSE): 139481179.931558
Mean Absolute Error (MAE): 5998.709521111241

Model: SVR
Mean Squared Error (MSE): 2477726085.337769
Mean Absolute Error (MAE): 22656.695548414456



The Random Forest Regressor seems to perform better than the other models. It has a lower MSE and MAE, indicating that, on average, it predicts the car prices with smaller errors compared to the other models.

Fine Tuning

The selcted random forest regressor is tuned to find optimal hyper parameters for the model using gridsearch cv

In [ ]:
from sklearn.model_selection import GridSearchCV


model = RandomForestRegressor(random_state=42)

# Define the parameter grid for GridSearchCV
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10]
}

# best combination of hyperparameters using gridsearch
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)
grid_search.fit(X_train, y_train)

# best model and its hyperparameters
best_model = grid_search.best_estimator_
best_params = grid_search.best_params_

# Evaluate on test set
y_pred = best_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print("Best Hyperparameters:", best_params)
print("Mean Squared Error (MSE):", mse)
print("Mean Absolute Error (MAE):", mae)

Actual vs Predicted (model)

In [30]:
model = RandomForestRegressor()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
y_pred_rounded = [round(value, 1) for value in y_pred]

#DataFrame for actual vs. predicted values
comparison_df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred_rounded})
print(comparison_df.head())

    Actual  Predicted
0  29695.0    33184.1
1  30495.0    26565.6
2  37650.0    36966.8
3  16170.0    17775.6
4   2000.0     2001.2
